### PR TITLE
remove version.h and define VERSION in CPPFLAGS instead

### PR DESCRIPTION
**Pull request by N-R-K**
Closes: https://github.com/nsxiv/nsxiv/issues/237

---

Didn't test too thoroughly but I think this should do the job.

### CODE REVIEW

**Comment by N-R-K** [comment]
@mamg22 does this fix the issue on your end?

**Comment by mamg22** [comment]
Yes, it fixes the issue.

**Comment by mamg22** [comment]
Was looking for a way to avoid calling `git` on every compiler invocation by defining the variable once in the makefile, but there doesn't seem to be a POSIX way of doing it yet.

But I believe it might be possible in future versions, there's [this proposal for assigning variable from shell command via `!=`](https://www.austingroupbugs.net/view.php?id=337).

Other than that, this works and seems fine to me.

**Review by mamg22 — APPROVED** [review]

**Review by TAAPArthur — APPROVED** [review]

**Comment by N-R-K** [comment]
Hmm, this is actually not correct. If someone compiles via `make` then let's say edits `main.c` makes a new commit and then recompiles, then it's not going to trigger option.o rebuild and `nsxiv -v` will show wrong version/commit-hash.

**Comment by N-R-K** [comment]
https://github.com/nsxiv/nsxiv/pull/238/commits/ab594a063898b16b084f40365a016ff25f8fda63 should do the trick I think.

**Comment by mamg22** [comment]
> [ab594a0](https://github.com/nsxiv/nsxiv/commit/ab594a063898b16b084f40365a016ff25f8fda63) should do the trick I think.

Yup, it works.

**Comment by N-R-K** [comment]
Think it makes more sense to do this like this instead: https://github.com/nsxiv/nsxiv/pull/238/commits/1a7ad3bce0af1592e258b667edeeb5ca6aad4c70

**Comment by N-R-K** [comment]
> Was looking for a way to avoid calling `git` on every compiler invocation by defining the variable once in the makefile, but there doesn't seem to be a POSIX way of doing it yet.

Only `options.c` needs it, so we could do it like this:

```diff
diff --git a/Makefile b/Makefile
index 202772d..7ea492b 100644
--- a/Makefile
+++ b/Makefile
@@ -30,7 +30,6 @@ inc_fonts_0 =
 inc_fonts_1 = -I/usr/include/freetype2 -I$(PREFIX)/include/freetype2
 
 CPPFLAGS = -D_XOPEN_SOURCE=700 \
-  -DVERSION=\""$$(git describe 2>/dev/null || printf '%s' "$(VERSION)")"\" \
   -DHAVE_LIBGIF=$(HAVE_LIBGIF) -DHAVE_LIBEXIF=$(HAVE_LIBEXIF) \
   -DHAVE_LIBWEBP=$(HAVE_LIBWEBP) -DHAVE_LIBFONTS=$(HAVE_LIBFONTS) \
   $(inc_fonts_$(HAVE_LIBFONTS))
@@ -67,9 +66,14 @@ nsxiv: $(OBJS)
 	$(CC) $(CFLAGS) $(CPPFLAGS) -c -o $@ $<
 
 $(OBJS): Makefile nsxiv.h config.h commands.h
-options.o: $(OBJS:.o=.c)
 window.o: icon/data.h
 
+options.o: $(OBJS:.o=.c)
+	@echo "CC $@"
+	$(CC) $(CFLAGS) $(CPPFLAGS) \
+	-DVERSION=\""$$(git describe 2>/dev/null || printf '%s' "$(VERSION)")"\" \
+	-c -o $@ options.c
+
 config.h:
 	@echo "GEN $@"
 	cp config.def.h $@
```

Not sure if it's worth it though.

**Comment by N-R-K** [comment]
TBH, I think I'd be fine with just leaving this be and let packagers deal with it via patching out `all` prerequisite from `install:` . But the current PR is fine by me as well since rebuilding options.c and calling `git describe` aren't that big of a deal.

ping: @mamg22 @WitherCubes @0mp @xfgusta

**Comment by 0mp** [comment]
It would be nice if the value of VERSION is checked first and only if it's not empty, git-describe is used.

**Comment by N-R-K** [comment]
> It would be nice if the value of VERSION is checked first and only if it's not empty, git-describe is used.

That defeats the point of git versioning. We want to use the git version if it exist and use VERSION as a fallback. This means if someone is on git version, `nsxiv -v` will show the git commit hash.

**Comment by mamg22** [comment]
> TBH, I think I'd be fine with just leaving this be and let packagers deal with it via patching out `all` prerequisite from `install:` . But the current PR is fine by me as well since rebuilding options.c and calling `git describe` aren't that big of a deal.

I'm fine with either approach. The first one could be added to a "for packagers" section somewhere near the installation section.

That said, I slightly prefer the approach in this PR.

**Comment by N-R-K** [comment]
> The first one could be added to a "for packagers" section somewhere near the installation section.

We could add it to known issues: #242.

I guess there's one other option that I haven't thought of: simply creating a dummy `.git/index` file if it doesn't exist. Not sure if there's any downside of this. Pushed the commit for review purposes: https://github.com/nsxiv/nsxiv/pull/238/commits/dd6c6880b198b7d824a72c74b72bd113ecf50f56

File: Makefile [hunk 1]
```diff
@@ -79,6 +79,7 @@ version.h: Makefile .git/index
 	echo "#define VERSION \"$${v:-$(VERSION)}\"" >$@
 
 .git/index:
+	if [ ! -d .git ] || [ ! -f $@ ]; then mkdir .git && touch $@ ; fi
```
Comment (mamg22) [inline Makefile:82]:
```suggestion
	mkdir -p .git && touch $@
```

Could it be simplified to this?

**Comment by mamg22** [comment]
Looks good to me.

**Comment by N-R-K** [comment]
Re-requesting review since the patch changed.

**Comment by TAAPArthur** [comment]
Feels a little dirty to me to create `.git`. Feel like it isn't uncommon to use that directory's existence as a quick check to see if one is in a top level git repo.

**Comment by TAAPArthur** [comment]
> > Feels a little dirty to me to create `.git`.
> 
> Any other alternative then? If not then we can probably just list it on Known Issues and let packagers deal with it.

None that haven't already been suggested. Personally, having the version being the last major version and ignoring git hashes it perfectly fine. Only updating the version when options.c changes is also fine. If one cared, they could just always `make clean; make`. So I think any of the earlier suggestions is better than a dummy .gitignore.

**Pull request closed without merging**